File: QuickSort/QuickSort/QuickSort.cpp
#include <iostream>
#include <vector>

using namespace std;

void quickSort(int* tab, int left, int right)
{
    if (right <= left) return;

    int i = left - 1;
    int j = right + 1;

    int pivot = tab[(left + right) / 2];

    while (1)
    {
        while (pivot > tab[++i]);

        while (pivot < tab[--j]);

        if (i <= j)
            swap(tab[i], tab[j]);
        else break;
    }

    if (left < j)
        quickSort(tab, left, j);

    if (right > i)
        quickSort(tab, i, right);
}


int main()
{
    int *tab, size;

    cout << "Insert the size of array" << endl;
    cin >> size;

    tab = new int[size];

    for (int i = 0; i < size; i++)
    {
        cout << "Insert the number for " << i + 1 << " window in the array" << endl;
        cin >> tab[i];
    }

    quickSort(tab, 0, size - 1);

    for (int i = 0; i < size; i++)
        cout << tab[i] << " ";

    cin.ignore();
    cin.get();
    return 0;
}

File: QuickSort.py
#Insert items to the array, notifying the user if the inserted value is "" or not a number. User can choose to quit Inserting numbers
def InsertNumbers():
    list_of_numbers = []
    iterator = 0

    while True:
        print(f"You are currently adding number {iterator + 1}")
        print("To exit type: 'q'")
        number = input("Insert number: ")

        if number.strip() == "":
            print("Input cannot be empty. Please enter a number.")
            continue
        if number.lower() == "q":
            break

        try:
            number = float(number)
            list_of_numbers.append(number)
            iterator += 1

        except ValueError:
            print("Please insert a proper number.")
            continue

    print(f"You've added {iterator} items")
    print(list_of_numbers)

    return list_of_numbers

#Dont sort if the length of the array <= 1. Else divide array into 3 new arrays by using List Comprehensions
def QuickSort(numbers):
    length = len(numbers)
    if length <= 1:
        return numbers
    
    pivot = numbers[length // 2]
    left = [x for x in numbers if x < pivot]
    middle = [x for x in numbers if x == pivot]
    right = [x for x in numbers if x > pivot]

    return QuickSort(left) + middle + QuickSort(right)

def main():
    list_of_numbers = InsertNumbers()
    if list_of_numbers:
        sorted_numbers = QuickSort(list_of_numbers)
        print(f"Sorted array:  {sorted_numbers}")
    else:
        print("Theres nothing to sort.")


main()
